python/src/iceberg/avro/resolver.py

```diff
@@ -0,0 +1,150 @@
+
+     Args:
+         write_schema (Schema | IcebergType): The write schema of the Avro file
+         read_schema (Schema | IcebergType): The requested read schema which is equal or a subset of the write schema
+
+     Raises:
+         NotImplementedError: If attempting to resolve an unrecognized object type
+    """
+    raise NotImplementedError("Cannot resolve non-type: %s" % write_schema)
+
+
+@resolve.register(Schema)
+def _(write_schema: Schema, read_schema: Schema) -> Reader:
+    """Visit a Schema and starts resolving it by converting it to a struct"""
+    return resolve(write_schema.as_struct(), read_schema.as_struct())
+
+
+@resolve.register(StructType)
+def _(write_struct: StructType, read_struct: StructType) -> Reader:
+    """Iterates over the write schema, and checks if the field is in the read schema"""
+    results: List[Tuple[Optional[int], Reader]] = []
+
+    read_fields = {field.name: (pos, field) for pos, field in enumerate(read_struct.fields)}
+
+    for write_field in write_struct.fields:
+        if write_field.name in read_fields:
+            read_pos, read_field = read_fields[write_field.name]
+            result_reader = resolve(write_field.field_type, read_field.field_type)
+        else:
+            read_pos = None
+            result_reader = visit(write_field.field_type, ConstructReader())
+        result_reader = result_reader if write_field.required else OptionReader(result_reader)
+        results.append((read_pos, result_reader))
+
+    write_fields = {field.name: field for field in write_struct.fields}
+    for pos, read_field in enumerate(read_struct.fields):
+        if read_field.name not in write_fields:
+            if read_field.required:
+                raise ResolveException(f"{read_field} is in not in the write schema, and is required")
+            # Just set the new field to None
+            results.append((pos, NoneReader()))
+
+    return StructReader(tuple(results))
+
+
+@resolve.register(ListType)
+def _(write_list: ListType, read_list: ListType) -> Reader:
+    if not isinstance(read_list, ListType):
+        raise ResolveException(f"Cannot change {write_list} into {read_list}")
+    element_reader = resolve(write_list.element.field_type, read_list.element.field_type)
+    return ListReader(element_reader)
+
+
+@resolve.register(MapType)
+def _(write_map: MapType, read_map: MapType) -> Reader:
+    if not isinstance(read_map, MapType):
+        raise ResolveException(f"Cannot change {write_map} into {read_map}")
+    key_reader = resolve(write_map.key.field_type, read_map.key.field_type)
+    value_reader = resolve(write_map.value.field_type, read_map.value.field_type)
+
+    return MapReader(key_reader, value_reader)
+
+
+ALLOWED_PROMOTIONS: Dict[Type[PrimitiveType], Set[Type[PrimitiveType]]] = {
+    # For now we only support the binary compatible ones
+    IntegerType: {LongType},
+    StringType: {BinaryType},
+    BinaryType: {StringType},
```

Comment:
I don't think this is currently allowed by the spec, but I don't see a reason why it shouldn't be. Maybe we should add it to the spec? What do you think?

Reply:
For background, one of the restrictions that we place on type promotion is that promotion is only allowed if hashing the value produces the same result. That's why `hashInt(v)` is implemented as `hashLong(castToLong(v))`. Otherwise, when a table column is promoted from `int` to `long`, any metadata values for `bucket` partitions would be suddenly incorrect.

It should be possible to promote from `binary` to `string` (or the opposite) because the hash value is the same.

I think we can also relax this constraint a bit, so if there is no bucket transform on a column, you can perform a type promotion that would not be allowed otherwise. For example, int to string promotion could only be allowed if there was no partition spec with a bucket function on the int column.

There are also odd cases with timestamps. If a long value is a timestamp in microseconds, then it could be promoted to `timestamp` or `timestamptz` because the hash function would match (no need to modify the value). But if the long was a timestamp in milliseconds, we could only promote to `timestamp` or `timestamptz` if there was no bucket transform applied to the value in a partition spec.

Reply:
I got this from the Avro schema evolution:

![Screenshot 2022-06-28 at 08 48 18](https://user-images.githubusercontent.com/1134248/176112107-fbe204bc-896a-4c7d-b2a5-b0d43eb8595b.png)

And thanks for the elaborate background information. And I agree that we should add it to the spec indeed. For the hashing, for Python it seems to check out:
```python
➜  python git:(fd-resolve-write-read-schemas) ✗ python3
Python 3.9.13 (main, May 24 2022, 21:13:51) 
[Clang 13.1.6 (clang-1316.0.21.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> hash("vo")
292714697147949111
>>> hash("vo".encode("utf8"))
292714697147949111
```

For Java, this does not seem to be the case:
```
➜  python git:(fd-resolve-write-read-schemas) ✗ scala
Welcome to Scala 2.13.8 (OpenJDK 64-Bit Server VM, Java 18.0.1.1).
Type in expressions for evaluation. Or try :help.

scala> new java.lang.String("vo").hashCode
val res1: Int = 3769

scala> new java.lang.String("vo").getBytes()
val res2: Array[Byte] = Array(118, 111)

scala> new java.lang.String("vo").getBytes().hashCode()
val res3: Int = 1479286669
```

We should also make sure that the hashes are the same across languages. I think this is okay for now because we implement this using the MurmurHash library.

Reply:
Hashing in this case is documented in the Iceberg spec: https://iceberg.apache.org/spec/#appendix-b-32-bit-hash-requirements

That's the one I'm talking about, and we have verified that it is the same between Java and Python.

I also want to note that the type promotions that we care about are the promotions that are allowed in Iceberg, not Avro. While we're using Avro files, Avro is more permissive in these cases than Iceberg allows.

Reply:
I've created a PR here to add string/binary conversion to the Spec: https://github.com/apache/iceberg/pull/5151

> I also want to note that the type promotions that we care about are the promotions that are allowed in Iceberg, not Avro. While we're using Avro files, Avro is more permissive in these cases than Iceberg allows.

Good point, I was focussing on the Avro ones here (since it is in the Avro package), but we can also make this pluggable by providing different promotion dictionaries for Avro and Iceberg.